test/org/sagebionetworks/bridge/dynamodb/DynamoExternalIdDaoMockTest.java

```diff
@@ -1,87 +1,252 @@
+
+        DynamoDBQueryExpression<DynamoExternalIdentifier> query = queryCaptor.getValue();
+
+        Condition rangeCondition = query.getRangeKeyConditions().get(DynamoExternalIdDao.IDENTIFIER);
+        assertEquals(BEGINS_WITH.toString(), rangeCondition.getComparisonOperator());
+        assertEquals("foo", rangeCondition.getAttributeValueList().get(0).getS());
 
-        when(mapper.queryPage(eq(DynamoExternalIdentifier.class), any())).thenReturn(resultPage);
+        Condition assignmentCondition = query.getQueryFilter().get(DynamoExternalIdDao.HEALTH_CODE);
+        assertEquals(NOT_NULL.toString(), assignmentCondition.getComparisonOperator());
 
-        ForwardCursorPagedResourceList<ExternalIdentifierInfo> result =
-                dao.getExternalIds(
-                        new StudyIdentifierImpl("studyId"),
-                        null,
-                        pageSize,
-                        null,
-                        false);
+        // This has been nullified, despite the fact that we provided an offset key
+        assertNull(query.getExclusiveStartKey());
+    }
+
+    @Test
+    public void getExternalIdsFiltersSubstudyInExternalIdentifierInfo() {
+        BridgeUtils.setRequestContext(new RequestContext.Builder()
+                .withCallerSubstudies(ImmutableSet.of("substudyA", "substudyB")).build());
+        
+        // Verify here that prior to migration, a lack of association doesn't break anything
+        DynamoExternalIdentifier extId1 = new DynamoExternalIdentifier(TestConstants.TEST_STUDY_IDENTIFIER, "extId1");
+        extId1.setSubstudyId(null); // Should see this substudy
+        DynamoExternalIdentifier extId2 = new DynamoExternalIdentifier(TestConstants.TEST_STUDY_IDENTIFIER, "extId2");
+        extId2.setSubstudyId("substudyA");
+        DynamoExternalIdentifier extId3 = new DynamoExternalIdentifier(TestConstants.TEST_STUDY_IDENTIFIER, "extId3");
+        extId3.setSubstudyId("substudyB");
+        DynamoExternalIdentifier extId4 = new DynamoExternalIdentifier(TestConstants.TEST_STUDY_IDENTIFIER, "extId4");
+        extId4.setSubstudyId("substudyC");
+        DynamoExternalIdentifier extId5 = new DynamoExternalIdentifier(TestConstants.TEST_STUDY_IDENTIFIER, "extId5");
+        extId5.setSubstudyId("substudyD");
+        setupQueryOfIds(ImmutableList.of(extId1, extId2, extId3, extId4, extId5));
+        
+        ForwardCursorPagedResourceList<ExternalIdentifierInfo> externalIds = dao.getExternalIds(TEST_STUDY, null, 10,
+                null, null);
+        assertNull(externalIds.getItems().get(0).getSubstudyId());
+        assertEquals("substudyA", externalIds.getItems().get(1).getSubstudyId());
+        assertEquals("substudyB", externalIds.getItems().get(2).getSubstudyId());
+        assertNull(externalIds.getItems().get(3).getSubstudyId());
+        assertNull(externalIds.getItems().get(4).getSubstudyId());
```

Comment:
Should we verify that we can see all of the external IDs, even if some of the substudies have been filtered out?

Reply:
If the substudy is filtered out then you will not see the external ID associated with that substudy. Sage admins will continue to see everything associated to the account because we're not associated to one or more substudies.